pdfminer/ccitt.py

```diff
@@ -521,61 +540,63 @@ def _do_uncompressed(self, bits):
 
 class CCITTFaxDecoder(CCITTG4Parser):
 
-    def __init__(self, width, bytealign=False, reversed=False):
+    def __init__(self, width: int, bytealign: bool = False,
+                 reversed: bool = False):
         CCITTG4Parser.__init__(self, width, bytealign=bytealign)
         self.reversed = reversed
         self._buf = b''
         return
 
-    def close(self):
+    def close(self) -> bytes:
         return self._buf
 
-    def output_line(self, y, bits):
+    def output_line(self, y: int, bits: Sequence[int]) -> None:
         bytes = array.array('B', [0]*((len(bits)+7)//8))
         if self.reversed:
             bits = [1-b for b in bits]
         for (i, b) in enumerate(bits):
             if b:
                 bytes[i//8] += (128, 64, 32, 16, 8, 4, 2, 1)[i % 8]
-        self._buf += bytes.tostring()
+        self._buf += bytes.tobytes()
         return
 
 
-def ccittfaxdecode(data, params):
+def ccittfaxdecode(data: bytes, params: Dict[str, object]) -> bytes:
     K = params.get('K')
-    cols = params.get('Columns')
-    bytealign = params.get('EncodedByteAlign')
-    reversed = params.get('BlackIs1')
+    cols = cast(int, params.get('Columns'))
+    bytealign = cast(bool, params.get('EncodedByteAlign'))
+    reversed = cast(bool, params.get('BlackIs1'))
```

Comment:
This interprets the memory that contains the parameter as another type. That will fix the type errors but might introduce other errors. I expect it is better to explicitly convert the parameter to the type.

E.g. do `int(params.get('Columns'))` instead of `cast(int, params.get('Columns'))`. If the value of the parameter is `"100"`, this will work. 

Curious to know if you had specific reasons to use `cast()` instead.

Reply:
My general strategy was trying not to break existing code, so I only introduced a runtime cast or assertion when the code was clearly broken. In this case, the callee (CCITTFaxDecoder) assumes that cols is going to be a bare (not-None) int, but we have no useful type information for params. Assuming that the code works, then the parameter is already int at runtime and a runtime conversion would be either redundant or worse potentially incorrect. I figured having the casts in the codebase was a better approach, and one we can try to drive down over time with stricter typing (e.g. less use of object or Any).

This kind of problem shows up wherever we have (a) objects on the PS/PDF stack, and (b) general dicts like attrs and params.

Reply:
I was also trying to be very careful about potentially changing runtime behaviour, just because this PR is so big already. Certainly reducing the number of casts and type: ignore comments is a good goal, but it may make more sense to make changes like this once the types are in.

Reply:
Ah, I see. I was looking at python [cast](https://docs.python.org/3.8/library/stdtypes.html#memoryview.cast) instead of typing [cast](https://docs.python.org/3/library/typing.html#typing.cast). 

Being careful about changing runtime behavior is a very good thing :+1: